File: split_array_largest_sum.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
class Solution {
    bool ok(int arr[], int N, int k , int mi) {
        int t = 0;
        for(int i = 0; i < N; ++i) {
            t += arr[i];
            if (t > mi) {
                if (--k == 0) return false;
                t = arr[i];
            }
        }
        return true;
    }
  public:
    int splitArray(int arr[] ,int N, int K) {
        // code here
        int hi = accumulate(arr, arr+N, 0);
        int lo = *max_element(arr, arr+N);
        int ans = -1;
        while(lo <= hi) {
            int mi = lo + (hi - lo) / 2;
            if (ok(arr, N, K, mi)) {
                ans = mi;
                hi = mi - 1;
            } else {
                lo = mi + 1;
            }
        }
        return ans;
    }
};

//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {
        int N, K;
        
        cin>>N>>K;
        int arr[N];
        
        for(int i=0 ; i<N ; i++)
            cin>>arr[i];

        Solution ob;
        cout<<ob.splitArray(arr,N,K);
        cout<<"\n";
    }
    return 0;
}
// } Driver Code Ends

File: binary_search/split_array_largest_sum.py
# User function Template for python3


class Solution:
    def splitArray(self, arr, N, K):
        # code here
        def ok(v):
            curr = 0
            cnt = 0
            for a in arr:
                if curr + a > v:
                    cnt += 1
                    curr = a
                else:
                    curr += a
            if curr > 0:
                cnt += 1

            return cnt <= K

        lo, hi = max(arr), sum(arr)
        ans = hi
        while lo <= hi:
            mid = (lo + hi) // 2
            if ok(mid):
                ans = mid
                hi = mid - 1
            else:
                lo = mid + 1

        return ans


# {
# Driver Code Starts
# Initial Template for Python 3

if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        N, K = map(int, input().split())
        arr = list(map(int, input().split()))

        ob = Solution()
        print(ob.splitArray(arr, N, K))
# } Driver Code Ends
